### PR TITLE
chore(make): handles kustomize substitutions without polluting history

### PR DESCRIPTION
**Pull request by bartoszmajsak**
## Description

- introduces a wrapper to call `kustomize` with customizations and undo after
- simplifies make targets
- removes the need for `.yaml.in` convention and enables kustomize manifests as a whole can be validated and supported by IDEs
- sets a reference to existing manager image instead of a placeholder; this makes kustomize work without Makefile involvement

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by ykaliuta** [comment]
For me the header does not sound corresponding reality since it does not pollute any history as is.

There are some drawbacks:
- any uncommitted change to the file will be silently reverted;
- after execution on the filesystem there is no snapshot of exact kustomize files applied to the cluster.

I'm not fan of this solution either, but accept the compromise.

File: config/manager/kustomization.yaml [hunk 1]
```diff
@@ -8,8 +10,8 @@ configMapGenerator:
 - files:
   - controller_manager_config.yaml
   name: manager-config
-apiVersion: kustomize.config.k8s.io/v1beta1
-kind: Kustomization
+
 images:
 - name: controller
-  newName: REPLACE_IMAGE
+  newName: quay.io/opendatahub/opendatahub-operator
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply config/manager/kustomization.yaml:16]: if i recall this correctly, at first we put REPLACE_IMAGE here is for openshift-release to work with image name.

File: Makefile [hunk 2]
```diff
@@ -186,30 +186,32 @@ ifndef ignore-not-found
...
+prepare: manifests kustomize
+
+# - makes necessary substitutions in kustomize manifests
+# - calls actual kustomize command
+# - restores manifests to original state
+define kustomize-with-undo
+trap 'git restore config/manager/kustomization.yaml' EXIT
+$(shell cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG))
+$(KUSTOMIZE) $(1)
+endef
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply Makefile:198]: Have you tried it?

File: config/manager/kustomization.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply config/manager/kustomization.yaml:16]: > it's a side change to the main goal of the PR.

It's not a side change. It makes the whole thing working out of the box instead. Previously it always required extra steps, either running `kustomize edit` or running make target to create non-existing `kustomize.yaml` before you could actually use kustomize.

File: Makefile [hunk 3]
```diff
@@ -186,30 +186,32 @@ ifndef ignore-not-found
...
-.PHONY: manager-kustomization
-manager-kustomization: config/manager/kustomization.yaml.in
-	cd config/manager \
-		&& cp -f kustomization.yaml.in kustomization.yaml \
-		&& $(KUSTOMIZE) edit set image controller=$(IMG)
+prepare: manifests kustomize
+
+# - makes necessary substitutions in kustomize manifests
+# - calls actual kustomize command
+# - restores manifests to original state
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply Makefile:193]: The intent here was to indicate that if you want to use more `kustomize edit`s make it here. I can either reword it to capture it or remove it. As you wish.

**Comment by bartoszmajsak** [comment]
> For me the header does not sound corresponding reality since it does not pollute any history as is.

Open to suggestions here. Prior to `yaml.in` approach it was leaving the tracked file with changes imposing the risk of polluting history with references to custom image. With the latest changes, it generates an untracked file and introduces a risk of accidentally committing it.

> There are some drawbacks:
> 
> * any uncommitted change to the file will be silently reverted;

This file in my eyes is rather stable and if you need to change some things in it you will pay attention to make sure it will be committed in a granular PR. Notice that the undo part is only done for `(un)deploy` and `(un)install` targets which are anyway targeted for local dev aka inner loop. But perhaps instead of doing `git restore` we can make undo a bit more accurate.

> * after execution on the filesystem there is no snapshot of exact kustomize files applied to the cluster.

Is there a need for that? Applying kustomize and piping it to `kubectl apply` anyway results in losing this information, as actual manifests are generated on the fly and output to stdout.

 > I'm not fan of this solution either, but accept the compromise.

The current state of this is also far from ideal, so let's work together to make it better instead of accepting compromises.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ykaliuta) [reply Makefile:198]: > Have you tried it?

yes

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply Makefile:198]: I think I tricked myself here :) Fixed in 37b1306

File: config/manager/kustomization.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply config/manager/kustomization.yaml:16]: The image already exists...

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ykaliuta) [reply Makefile:194]: Why do you need trap? If you afraid of any command to fail, you can add - in front of or ||: at the end.

**Comment by bartoszmajsak** [comment]
I think the debate to "lines of code ratio" is far beyond reasonable proportion here and goes into an unhealthy nitpicking direction instead of trying to improve far from ideal situation.

Closing in favor of the current state of things.

**Comment by ykaliuta** [comment]
> I think the debate to "lines of code ratio" is far beyond reasonable proportion here and goes into an unhealthy nitpicking direction instead of trying to improve "far from ideal" situation.
> 
> Closing in favor of the current state of things.

Hmm, sounds very emotional. If you are talking about me, than, as I said, I perfectly accept the solution if it's better for you and others. The only real concern was about overcomplicated implementation and I asked for the reasons. Others are not objections but just a conversation.

File: config/manager/kustomization.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply config/manager/kustomization.yaml:16]: > if i recall this correctly, at first we put REPLACE_IMAGE here is for openshift-release to work with image name.

so with the `yaml.in` file - how is it supposed to work now @zdtsw?

**Comment by bartoszmajsak** [comment]
It's not emotional. This is not an essential change and we can survive with the current state of things, so I took liberty of closing it.

But I think one aspect brought by @zdtsw should be verified https://github.com/opendatahub-io/opendatahub-operator/pull/836#discussion_r1476037662

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply Makefile:194]: Ah ok, I see what you mean. Sure, this can be done like that too. Wondering why we didn't start the whole convo with this...

[the remark this answers is not in the document]
Comment (ykaliuta) [reply Makefile:194]: > If `kustomize apply` succeeds adding `||` at the end won't have an effect, would it? And I wanted to restore regardless of the state.
So I guessed right. Yes, `|| :` or `|| true` or even just `; true`. `:` is NOP operator in shell.  Or make's minus https://www.gnu.org/software/make/manual/html_node/Errors.html

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply Makefile:194]: Yes, I realized that. Thank you for the pointers though

**Comment by ykaliuta** [comment]
> > For me the header does not sound corresponding reality since it does not pollute any history as is.
> 
> Open to suggestions here. 

I do not insist at all, just something like `revert kustomize substitutions after applying` sounds more about what's going on.

> Prior to `yaml.in` approach it was leaving the tracked file with changes imposing the risk of polluting history with references to custom image. With the latest changes, it generates an untracked file and introduces a risk of accidentally committing it.

> 
> > There are some drawbacks:
> > 
> > * any uncommitted change to the file will be silently reverted;
> 
> This file in my eyes is rather stable and if you need to change some things in it you will pay attention to make sure it will be committed in a granular PR.

Sure, it's not objection. Just part of the 'pros and cons' list.

> Notice that the undo part is only done for `(un)deploy` and `(un)install` targets which are anyway targeted for local dev aka inner loop.

Yes, and from the previous conversation, what about bundle target then? It has the same problem IIRC.

> But perhaps instead of doing `git restore` we can make undo a bit more accurate.

No no no, it will be an overkill for this simple case

> 
> > * after execution on the filesystem there is no snapshot of exact kustomize files applied to the cluster.
> 
> Is there a need for that? Applying kustomize and piping it to `kubectl apply` anyway results in losing this information, as actual manifests are generated on the fly and output to stdout.
> 
> > I'm not fan of this solution either, but accept the compromise.
> 
> The current state of this is also far from ideal, so let's work together to make it better instead of accepting compromises.

I do not see ideal solution yet. But in general I can say I  prefer your way more then generation from .in
